### ISSUE TITLE
Useful JSDoc rules (JSCS and Others)

### ISSUE THREAD

**Issue by simonepri** [eslint/eslint#9827]
I have read the issue discussing about the [gaps between ESLint  and JSCS](https://github.com/eslint/eslint/issues/5856) but I haven't found nothing about the [JSDoc rules](http://jscs.info/rule/jsDoc).
I'm not using JSCS but they seems really useful.

ESLint now supports only these 2 rules:
https://eslint.org/docs/rules/require-jsdoc
https://eslint.org/docs/rules/valid-jsdoc

I've created a table to show some gaps between ESLint and JSCS.

JSCS Rule | ESLint Rule | External Rules | Status
---|---|---|---
[checkAnnotations](http://jscs.info/rule/jsDoc#checkAnnotations) | [valid-jsdoc](https://eslint.org/docs/rules/valid-jsdoc#prefer) Allows to set the prefer option, but does not report an error for non-existing annotations | | ⚠️ 
[checkParamExistence](http://jscs.info/rule/jsDoc#checkParamExistence) | [valid-jsdoc](https://eslint.org/docs/rules/valid-jsdoc#rule-details) | | ✅
[checkParamNames](http://jscs.info/rule/jsDoc#checkParamNames) | [valid-jsdoc](https://eslint.org/docs/rules/valid-jsdoc#rule-details) | | ✅
[requireParamTypes](http://jscs.info/rule/jsDoc#requireParamTypes) | [valid-jsdoc](https://eslint.org/docs/rules/valid-jsdoc#rule-details) | | ✅
[checkRedundantParams](http://jscs.info/rule/jsDoc#checkRedundantParams) | | | ❌
[checkReturnTypes](http://jscs.info/rule/jsDoc#checkReturnTypes) | | | ❌
[checkRedundantReturns](http://jscs.info/rule/jsDoc#checkRedundantReturns) | | | ❌
[requireReturnTypes](http://jscs.info/rule/jsDoc#requireReturnTypes) | [valid-jsdoc](https://eslint.org/docs/rules/valid-jsdoc#rule-details) | | ✅
[checkTypes](http://jscs.info/rule/jsDoc#checkTypes) | [valid-jsdoc](https://eslint.org/docs/rules/valid-jsdoc#prefertype) Allows to set the preferType option, but does not have defaults | | ⚠️
[checkRedundantAccess](http://jscs.info/rule/jsDoc#checkRedundantAccess) | | | ❌
[leadingUnderscoreAccess](http://jscs.info/rule/jsDoc#leadingUnderscoreAccess) | | | ❌
[enforceExistence](http://jscs.info/rule/jsDoc#enforceExistence) | [require-jsdoc](https://eslint.org/docs/rules/require-jsdoc) | | ✅
[requireHyphenBeforeDescription](http://jscs.info/rule/jsDoc#requireHyphenBeforeDescription) | | [jsdoc/require-hyphen-before-param-description](https://github.com/gajus/eslint-plugin-jsdoc#eslint-plugin-jsdoc-rules-require-hyphen-before-param-description) rule can be merged inside ESLint | ⚠️
[requireNewlineAfterDescription](http://jscs.info/rule/jsDoc#requireNewlineAfterDescription) | | [jsdoc/newline-after-description](https://github.com/gajus/eslint-plugin-jsdoc#eslint-plugin-jsdoc-rules-newline-after-description) rule can be merged inside ESLint | ⚠️
[disallowNewlineAfterDescription](http://jscs.info/rule/jsDoc#disallowNewlineAfterDescription) | | [jsdoc/newline-after-description](https://github.com/gajus/eslint-plugin-jsdoc#eslint-plugin-jsdoc-rules-newline-after-description) rule can be merged inside ESLint | ⚠️
[requireDescriptionCompleteSentence](http://jscs.info/rule/jsDoc#requireDescriptionCompleteSentence) | | [jsdoc/require-description-complete-sentence](https://github.com/gajus/eslint-plugin-jsdoc#eslint-plugin-jsdoc-rules-require-description-complete-sentence) rule can be merged inside ESLint | ⚠️
[requireParamDescription](http://jscs.info/rule/jsDoc#requireParamDescription) | [valid-jsdoc](https://eslint.org/docs/rules/valid-jsdoc#rule-details) | | ✅
[requireReturnDescription](http://jscs.info/rule/jsDoc#requireReturnDescription) | [valid-jsdoc](https://eslint.org/docs/rules/valid-jsdoc#rule-details) | | ✅

**Comment by simonepri** [comment]
Possible defaults for `preferType` can be either:
```js
"preferType": {
  "array": "Array",
  "Boolean": "boolean",
  "date": "Date",
  "Number": "number",
  "object": "Object",
  "regex": "RegExp",
  "String": "string"
}
```
or 
```js
"preferType": {
  "array": "Array",
  "boolean": "Boolean",
  "date": "Date",
  "number": "Number",
  "object": "Object",
  "regex": "RegExp",
  "string": "String"
}
```

**Comment by ilyavolodin** [comment]
@simonepri Thanks for the list. Have you looked at https://github.com/gajus/eslint-plugin-jsdoc ? Our goal with JSCS migration is not to have every single JSCS rule in ESLint, but to have a way of migrating out of JSCS. We specifically decided to drop some of the JSCS rules, because they were not generic enough (for example JSCS had a rule specifically for jQuery usage, ESLint doesn't allow framework specific rules).

**Comment by simonepri** [comment]
Hi @ilyavolodin!
Yes, actually the link you provided is the same you can find in the "External Rules" column if you click on one of them. The issue is more about the ESLint's support for JSDoc more than an issue about the JSCS migration. I just mention it because I found he has some nice rules for JSDoc.

Nevermind, looking at the table, for example, are these two rules:
`require-hyphen-before-param-description`
`newline-after-description`
not generic enough as you told?

It may make sense for you to have a new rule inside ESLint called `style-jsdoc` that enable users to deal with formatting rules for the JSDocs?
If so, I'm welcome to contribute to do so. But I prefer to discuss it here, before actually doing a PR.

Thanks!

**Comment by ilyavolodin** [comment]
I personally have a mixed feelings about adding more JSDoc rules to the core. On one hand - they are useful, and I use them myself. On the other - JSDoc is not part of ECMAScript spec at all, and is a convention that's built on top of JS. More than that, there doesn't even seems to be an agreement on the convention itself. Doctrine that ESLint uses, doesn't follow the same rules as Google's JSDoc parser. From my perspective, I would like to see those rules added, but maybe to eslint-plug-jsdoc instead of the core.

**Comment by simonepri** [comment]
I perfectly understand your point of view.
Actually I agree with the fact that an external plugin may seem a great choice.

But question arises naturally.
If that's your feeling, why there are those rules about JSDoc inside ESLint?

It does not makes too sense for me to have some rules inside ESLint and others inside a plugin.

My thought (maybe wrong) when I first saw those rules was:
"Mhh...cool they have rules for JSDoc let's enable them!"
"Umh..they are pretty limited."
"Ok, let's search for a plugin."
"Mhh...this plugin has rules that overlaps with the rules inside the ESLint core...which should I use?"

**Comment by ilyavolodin** [comment]
Both of the JSDoc rules that we have in the core has been added a long time ago, before ESLint had support for plugins (as far as I remember, at least). There was even a conversation at some point about moving them out into eslint-plugin-jsdoc. Now that we have a policy of not removing existing rules from the core, we wouldn't be able to get rid of them at all (just deprecate them, which feels somewhat useless, when we don't provide a new rule to replace deprecated one). I'm not strongly oppose to adding new JSDoc rules to the core if deemed necessary, but let's see how the rest of the team feels.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
